File: artemis-serialization/artemis-bytecode/src/main/java/com/artemis/compile/Node.java
package com.artemis.compile;

import java.util.ArrayList;
import java.util.List;

/**
 * <p>Data structure for mapping json into an intermediate
 * data structure, linking the hierarchy back to the original
 * java types.</p>
 */
public final class Node {
	public final Meta meta;
	public final Object payload;
	private final List<Node> children = new ArrayList<>();

	Node(Class<?> type, String field, Object payload) {
		meta = new Meta(type, field);
		this.payload = payload;
	}

	public List<Node> children() {
		return children;
	}

	public boolean add(Node node) {
		return children.add(node);
	}

	public static Node node(Class<?> type,
	                        Node... children) {

		Node n = new Node(type, null, null);
		for (Node child : children)
			n.children.add(child);

		return n;
	}

	public static Node node(Class<?> type,
	                        String field,
	                        Node... children) {

		return node(type, field, null, children);
	}

	public static Node node(Class<?> type,
	                        String field,
	                        Object payload,
	                        Node... children) {

		Node n = new Node(type, field, payload);
		for (Node child : children)
			n.children.add(child);

		return n;
	}

	private Object widenPayload() {
		Class<?> type = meta.type;
		if (payload instanceof Integer)
			return payload;
		if (type == Character.class || type == char.class)
			return new Integer(((Character) payload).charValue());
		if (type == Short.class || type == Byte.class)
			return Integer.parseInt(payload.toString());
		if (type == short.class || type == byte.class)
			return Integer.parseInt(payload.toString());

		return payload;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		Node node = (Node) o;

		if (!meta.equals(node.meta))
			return false;

		if (!children.equals(node.children))
			return false;

		return payload != null
			? widenPayload().equals(node.widenPayload())
			: node.payload == null;
	}

	@Override
	public int hashCode() {
		int result = meta.hashCode();
		result = 31 * result + children.hashCode();
		result = 31 * result + (payload != null ? payload.hashCode() : 0);
		return result;
	}

	private StringBuilder toStringBuilder(StringBuilder sb, String prepend) {
		String field = meta.field;
		String format =
			(field == null && payload == null) ? "(%2$s)" :
			(field == null && payload != null) ? "(%2$-10s %3$s)" :
		    (field != null && payload != null) ? "(%-10s %-10s %s)"
		                                       : "(%-10s %s)";

		format = prepend + format + '\n';
		sb.append(String.format(format, field, meta.type.getSimpleName(), payload));

		prepend += "    ";
		for (Node child : children)
			child.toStringBuilder(sb, prepend);

		return sb;
	}

	@Override
	public String toString() {
		return toStringBuilder(new StringBuilder(), "").toString();
	}

	public final static class Meta {
		public final Class<?> type;
		public final String field;

		public Meta(Class<?> type, String field) {
			this.type = type;
			this.field = field;
		}

		@Override
		public boolean equals(Object o) {
			if (this == o) return true;
			if (o == null || getClass() != o.getClass()) return false;

			Meta meta = (Meta) o;

			if (!type.equals(meta.type)) return false;
			return field != null ? field.equals(meta.field) : meta.field == null;
		}

		@Override
		public int hashCode() {
			int result = type.hashCode();
			result = 31 * result + (field != null ? field.hashCode() : 0);
			return result;
		}
	}
}

File: artemis-serialization/artemis-bytecode/src/main/kotlin/com/artemis/compile/node.kt
package com.artemis.compile

import com.badlogic.gdx.utils.JsonValue
import net.onedaybeard.transducers.Transducer
import net.onedaybeard.transducers.map
import net.onedaybeard.transducers.transduce
import kotlin.reflect.KClass

/**
 * Intermediate object graph representing mapped json and/or objects.
 */
data class Node(val type: Class<*> = Any::class.java,
                val field: String? = null,
                val payload: Any? = null,
                val children: MutableList<Node> = mutableListOf()) {

    constructor(type: KClass<*>,
                field: String? = null,
                payload: Any? = null,
                vararg nodes: Node) : this(type.java, field, payload) {

        children.addAll(nodes)
    }

    constructor(type: KClass<*>,
                field: String? = null,
                vararg nodes: Node) : this(type.java, field, null) {

        children.addAll(nodes)
    }

    constructor(type: KClass<*>, vararg nodes: Node) : this(type.java) {
        children.addAll(nodes)
    }

    fun add(node: Node): Boolean {
        return children.add(node)
    }

    operator fun get(index: Int): Node {
        return children[index]
    }

    override fun toString(): String {
        return format().toString()
    }

    private fun format(sb: StringBuilder = StringBuilder(), indent: String = ""): StringBuilder {
        sb.append("$indent${formatted()}\n")
        children.forEach { it.format(sb, "$indent    ") }

        return sb
    }

    private fun formatted(): String {
        val p = payload; val f = field

        val name = type.simpleName
        return if (f == null && p == null) "($name)"
          else if (f == null && p != null) "(${name.padEnd(10)} $p)"
          else if (f != null && p != null) "(${f.padEnd(10)} ${name.padEnd(10)} $p)"
          else                             "(${f!!.padEnd(10)} ${type.simpleName})"
    }
}


private val sanitizeNodes: Transducer<Node, Node> = map { n ->
    when (n.type.kotlin) {
        Short::class -> n.copy(payload = (n.payload as? Int)?.toShort() ?: n.payload)
        Byte::class  -> n.copy(payload = (n.payload as? Int)?.toByte()  ?: n.payload)
        else         -> n
    }
}

fun symbolToNode(owner: Any): Transducer<Node, Symbol> {
    return map { symbol: Symbol ->
        val value = owner.readField(symbol)
        if (symbol.isBuiltInType()) {
            Node(symbol.type, symbol.field, value)
        } else {
            toNode(value!!, symbol.field)
        }
    }
}

fun Any.readField(symbol: Symbol): Any? {
    return this.javaClass.getField(symbol.field).get(this)
}

fun toNode(obj: Any, name: String? = null): Node {
    return transduce(xf = symbolToNode(obj),
                     rf = buildNodeStep,
                     init = Node(type = obj.javaClass, field = name),
                     input = symbolsOf(obj))
}

fun toNode(type: Class<*>,
           json: JsonValue,
           name: String? = null): Node {

    return transduce(xf = symbolToNode(json) + sanitizeNodes,
                     rf = buildNodeStep,
                     init = Node(type = type, field = name),
                     input = symbolsOf(type))
}

fun Symbol.isBuiltInType(): Boolean = isBuiltInType(this.type)

fun isBuiltInType(t: Class<*>): Boolean = t.isPrimitive || t == String::class.java

private val buildNodeStep = { result: Node,
                              input: Node -> result.apply { add(input) } }
